flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/sink/TestSink.java

```diff
@@ -397,6 +400,54 @@ public void endOfInput() {
         }
     }
 
+    /**
+     * {@link GlobalCommitter} implementation that can throw on exception when processing
+     * checkpoint. Exception is thrown once.
+     */
+    public static class FailOnCommitGlobalCommitter extends DefaultGlobalCommitter {
+
+        private static final Logger LOG =
+                LoggerFactory.getLogger(FailOnCommitGlobalCommitter.class);
+
+        // local counter used to decide if exception should be thrown. This field has to be static,
+        // since we want to keep its value after Flink's local cluster recovery.
+        private static int globalCheckpointCounter = 0;
+
+        private final int commitNumberToFailOn;
+
+        /**
+         * Creates instance of FailOnCommitGlobalCommitter.
+         *
+         * @param checkpointNumberToFailOn number of checkpoints after which exception should be
+         *     thrown.
+         * @param queueSupplier queueSupplier.
+         */
+        public FailOnCommitGlobalCommitter(
+                int checkpointNumberToFailOn, Supplier<Queue<String>> queueSupplier) {
+            super(queueSupplier);
+            this.commitNumberToFailOn = checkpointNumberToFailOn;
+        }
+
+        @Override
+        public List<String> commit(List<String> committables) {
+            LOG.info(
+                    "Commit number "
+                            + globalCheckpointCounter
+                            + ", committables size "
+                            + committables.size());
+            try {
+                if (globalCheckpointCounter == commitNumberToFailOn) {
+                    throw new RuntimeException(
+                            "GlobalCommitter Desired Exception on checkpoint "
+                                    + globalCheckpointCounter);
+                }
+            } finally {
+                globalCheckpointCounter++;
```

Comment:
I think the code logic is a bit confusing because the exception happens one commit too late i.e.

globalCheckpointCounter, commitNumberToFailOn
0 -> 2
1 -> 2
2 -> 3

It takes 3 commits until the value is triggered. You can simply remove the `try` block and increment the global counter before the check.

Reply:
Thanks,
its changed now.